provisioning.go

```diff
@@ -0,0 +1,370 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"github.com/chrisccoulson/go-tpm2"
+
+	"golang.org/x/xerrors"
+)
+
+// ProvisionStatusAttributes correspond to the state of the TPM with regards to provisioning for full disk encryption.
+type ProvisionStatusAttributes int
+
+// ProvisionMode is used to control the behaviour of ProvisionTPM.
+type ProvisionMode int
+
+const (
+	ppiPath string = "/sys/class/tpm/tpm0/ppi/request" // Path for submitting PPI operation requests to firmware for the default TPM
+
+	// clearPPIRequest is the operation value for asking the firmware to clear the TPM, see section 9 of "TCG PC Client Platform Physical
+	// Presence Interface Specification", version 1.30, revision 00.52, 28 July 2015.
+	clearPPIRequest string = "5"
+
+	// DA lockout parameters.
+	maxTries        uint32 = 32
+	recoveryTime    uint32 = 7200
+	lockoutRecovery uint32 = 86400
+)
+
+const (
+	// AttrValidSRK indicates that the TPM contains a valid primary storage key with the expected properties at the
+	// expected location. Note that this does not mean that the object was created with the same template that ProvisionTPM
+	// uses, and is no guarantee that a call to ProvisionTPM wouldn't result in a different key being created.
+	AttrValidSRK ProvisionStatusAttributes = 1 << iota
+
+	// AttrValidEK indicates that the TPM contains a valid endorsement key at the expected location. On a TPMConnection created
+	// with SecureConnectToDefaultTPM, it means that the TPM contains the key associated with the verified endorsment certificate.
+	// On a TPMConnection created with ConnectToDefaultTPM, it means that the TPM contains a valid primary key with the expected
+	// properties at the expected location, but does not mean that the object was created with the the same template that
+	// ProvisionTPM uses, and is no guarantee that a call to ProvisionTPM wouldn't result in a different key being created.
+	AttrValidEK
+
+	AttrDAParamsOK         // The dictionary attack lockout parameters are configured correctly.
+	AttrOwnerClearDisabled // The ability to clear the TPM with owner authorization is disabled.
+
+	// AttrLockoutAuthSet indicates that the lockout hierarchy has an authorization value defined. This
+	// doesn't necessarily mean that the authorization value is the same one that was originally provided
+	// to ProvisionTPM - it could have been changed outside of our control.
+	AttrLockoutAuthSet
+
+	AttrLockNVIndex // The TPM has a valid NV index used for locking access to keys sealed with SealKeyToTPM
+)
+
+const (
+	// ProvisionModeClear specifies that the TPM should be fully provisioned after clearing it.
+	ProvisionModeClear ProvisionMode = iota
+
+	// ProvisionModeWithoutLockout specifies that the TPM should be refreshed without performing
+	// operations that require knowledge of the lockout hierarchy authorization value.
```

Comment:
this should probably expand a bit on what those operations are,  if I understand, they are  disabling owner clear and configuring lockout itself ?

Reply:
Done